x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/metadata/MetadataMigrateToDataTiersRoutingService.java

```diff
@@ -0,0 +1,309 @@
+                currentMetadata.getOperationMode() + "]");
+        }
+
+        Metadata.Builder mb = Metadata.builder(currentState.metadata());
+        String removedIndexTemplateName = null;
+        if (Strings.hasText(indexTemplateToDelete) &&
+            currentState.metadata().getTemplates().containsKey(indexTemplateToDelete)) {
+            mb.removeTemplate(indexTemplateToDelete);
+            logger.debug("removing template [{}]", indexTemplateToDelete);
+            removedIndexTemplateName = indexTemplateToDelete;
+        }
+
+        String attribute = nodeAttrName;
+        if (Strings.isNullOrEmpty(nodeAttrName)) {
+            attribute = DEFAULT_NODE_ATTRIBUTE_NAME;
+        }
+        List<String> migratedPolicies = migrateIlmPolicies(mb, currentState, attribute, xContentRegistry, client);
+        List<String> migratedIndices = migrateIndices(mb, currentState, attribute);
+        return Tuple.tuple(ClusterState.builder(currentState).metadata(mb).build(),
+            new MigratedEntities(removedIndexTemplateName, migratedIndices, migratedPolicies));
+    }
+
+    static List<String> migrateIlmPolicies(Metadata.Builder mb, ClusterState currentState, String nodeAttrName,
+                                           NamedXContentRegistry xContentRegistry, Client client) {
+        IndexLifecycleMetadata currentLifecycleMetadata = currentState.metadata().custom(IndexLifecycleMetadata.TYPE);
+        if (currentLifecycleMetadata == null) {
+            return Collections.emptyList();
+        }
+
+        List<String> migratedPolicies = new ArrayList<>();
+        Map<String, LifecyclePolicyMetadata> currentPolicies = currentLifecycleMetadata.getPolicyMetadatas();
+        SortedMap<String, LifecyclePolicyMetadata> newPolicies = new TreeMap<>(currentPolicies);
+        for (Map.Entry<String, LifecyclePolicyMetadata> policyMetadataEntry : currentPolicies.entrySet()) {
+            LifecyclePolicy newLifecyclePolicy = migrateSingleILMPolicy(nodeAttrName, policyMetadataEntry.getValue().getPolicy());
+            if (newLifecyclePolicy != null) {
+                // we updated at least one phase
+                long nextVersion = policyMetadataEntry.getValue().getVersion() + 1L;
+                LifecyclePolicyMetadata newPolicyMetadata = new LifecyclePolicyMetadata(newLifecyclePolicy,
+                    policyMetadataEntry.getValue().getHeaders(), nextVersion, Instant.now().toEpochMilli());
+                LifecyclePolicyMetadata oldPolicyMetadata = newPolicies.put(policyMetadataEntry.getKey(), newPolicyMetadata);
+                assert oldPolicyMetadata != null :
+                    "we must only update policies, not create new ones, but " + policyMetadataEntry.getKey() + " didn't exist";
+
+                updateIndicesForPolicy(mb, currentState, xContentRegistry, client, oldPolicyMetadata.getPolicy(), newPolicyMetadata);
```

Comment:
Perhaps we should capture the return value for this so that we can have debug logging for the number of indices that had their cached policies updated?

Reply:
The return value is currently true/false - indicating if any index was updated (ie. if the `Metadata.Builder` is changed). I added the logging inside the method to keep the return value as is (not sure we need to expose more at this point)